linkerd/app/outbound/src/orig_proto_upgrade.rs

```diff
@@ -1,68 +1,72 @@
-use super::Endpoint;
+use crate::endpoint::Endpoint;
 use crate::proxy::http::{orig_proto, settings::Settings};
 use crate::svc;
 use futures::{try_ready, Future, Poll};
-use http;
-use std::marker::PhantomData;
 use tracing::trace;
 
-#[derive(Debug)]
-pub struct Layer<A, B>(PhantomData<fn(A) -> B>);
+#[derive(Clone, Debug, Default)]
+pub struct OrigProtoUpgradeLayer(());
 
-#[derive(Debug)]
-pub struct MakeSvc<M, A, B> {
+#[derive(Clone, Debug)]
+pub struct OrigProtoUpgrade<M> {
     inner: M,
-    _marker: PhantomData<fn(A) -> B>,
 }
 
-pub struct MakeFuture<F, A, B> {
+pub struct UpgradeFuture<F> {
     can_upgrade: bool,
     inner: F,
-    _marker: PhantomData<fn(A) -> B>,
+    was_absolute: bool,
 }
 
-pub fn layer<A, B>() -> Layer<A, B> {
-    Layer(PhantomData)
-}
-
-impl<A, B> Clone for Layer<A, B> {
-    fn clone(&self) -> Self {
-        Layer(PhantomData)
+impl OrigProtoUpgradeLayer {
+    pub fn new() -> Self {
+        Self::default()
     }
 }
 
-impl<M, A, B> svc::Layer<M> for Layer<A, B>
-where
-    M: svc::MakeService<Endpoint, http::Request<A>, Response = http::Response<B>>,
-{
-    type Service = MakeSvc<M, A, B>;
+impl<M> svc::Layer<M> for OrigProtoUpgradeLayer {
```

Comment:
Not a blocker, but any reason `app::core` rexports `tower::layer::Layer`? We could just use `tower::layer::Layer` direclty here and down below for `Either`. There's only a few other places we use the rexported types/traits.

Reply:
Classically, `svc` provided all of the service primitives to app::core. It doesn't really need to stay that way, though.